src/process/unix/SDL_process.c

```diff
@@ -0,0 +1,314 @@
+}
+
+SDL_Process *SDL_CreateProcess(const char * const *args, const char * const *env, SDL_ProcessFlags flags)
+{
+    // Keep the malloc() before exec() so that an OOM won't run a process at all
+    SDL_Process *process = SDL_malloc(sizeof(SDL_Process));
+
+    if (!process) {
+        return NULL;
+    }
+
+    process->flags = flags;
+
+    if ((flags & SDL_PROCESS_STDIN) && (pipe(process->stdin_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDOUT) && (pipe(process->stdout_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    if ((flags & SDL_PROCESS_STDERR) && (pipe(process->stderr_pipe) < 0)) {
+        SDL_SetError("Could not pipe(): %s", strerror(errno));
+
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+        return NULL;
+    }
+
+    process->pid = fork();
+
+    if (process->pid < 0) {
+        SDL_SetError("Could not fork(): %s", strerror(errno));
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[READ_END]);
+            close(process->stdin_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            close(process->stdout_pipe[WRITE_END]);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            close(process->stderr_pipe[WRITE_END]);
+        }
+
+        SDL_free(process);
+
+        return NULL;
+    } else if (process->pid == 0) {
+        if (flags & SDL_PROCESS_STDIN) {
+            close(process->stdin_pipe[WRITE_END]);
+            dup2(process->stdin_pipe[READ_END], STDIN_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDOUT) {
+            close(process->stdout_pipe[READ_END]);
+            dup2(process->stdout_pipe[WRITE_END], STDOUT_FILENO);
+        }
+
+        if (flags & SDL_PROCESS_STDERR) {
+            close(process->stderr_pipe[READ_END]);
+            dup2(process->stderr_pipe[WRITE_END], STDERR_FILENO);
+        }
+
+        char **mutable_args = dupstrlist(args);
+        char **mutable_env = env ? dupstrlist(env) : NULL;
+
+        // We are in a new process; don't bother freeing because either exec*() or exit() will succeed
+
+        if (!mutable_args || (env && !mutable_env)) {
+            fprintf(stderr, "Could not clone args/env: %s\n", SDL_GetError());
```

Comment:
`fprintf()` is not async-signal-safe. Doing I/O in an async-signal-safe location is, in general, a massive pain.

Reply:
On POSIX, I think I can just `write()` two times.

Reply:
Yes, if you assume that each `write()` will write out its entire argument and not be interrupted or a short write. (Pedantically, you need to `write()` in a loop, ignoring `EIO`, counting bytes written and updating a "cursor" pointing into the string, until everything has been written out - I told you it was a massive pain :-)

Is `SDL_GetError()` async-signal-safe? (Can it be implemented to be async-signal-safe, given how it works? I honestly don't know.)

Reply:
... three times, not two, in fact (prefix, error, newline).

Reply:
@madebr is currently writing a better implementation using SDL_Log. I'm not sure if SDL_Log is async-signal-safe, but I'll merge the changes shortly, so it would be good to check a second time for issues in that implementation.

Reply:
> Is SDL_GetError() async-signal-safe? 

In general, the answer to questions like this is:

* functions from libc are only async-signal-safe if `signal-safety(7)` specifically says so;
* functions from other libraries (including SDL!) are only async-signal-safe if their documentation specifically says so, so that their maintainers have a reminder to always keep async-signal-safety in mind when updating them

systemd's `sd_journal_stream_fd()` and GLib's `g_fdwalk_set_cloexec()` are among the rare examples of library functions that are documented to be async-signal-safe.

Reply:
> I'm not sure if SDL_Log is async-signal-safe

I would be amazed if it is.

Doesn't it call user-specified logging callbacks? If it does, it can't possibly be async-signal-safe (because the user-specified callback can't be guaranteed to be safe, and in fact almost certainly isn't).